### ISSUE TITLE
serverless deploy gets stuck forever after printing Endpoints

### ISSUE THREAD

**Issue by premraval-pr** [serverless/serverless#9555]
I have created a REST API using NodeJs and TypeScript using [aws-nodejs-typescript](https://github.com/serverless/serverless/tree/master/lib/plugins/create/templates/aws-nodejs-typescript) template. The whole code works fine on localhost using serverless-offline but as soon as I run serverless deploy, it starts all the steps, and then after printing endpoints, functions, and layers, it gets stuck. If I try to run the Endpoints given above, I get`{ "message": "Internal Server Error"}`  as a response to all POST, PATCH, and DELETE requests and all GET requests show some CloudFront error.

<details>
<summary><code>serverless.ts</code></summary>

```ts
import type { AWS } from "@serverless/typescript";
import {
	getUser,
	addUser,
	delUser,
	patchUser,
	loginUser,
	getUserByToken,
	getBuildings,
	addBuilding,
	delBuilding,
	patchBuilding,
	// getTemplates,
	addTemplate,
	delTemplate,
	patchTemplate,
	getUserLogs,
	getTemplateLogs,
	addUserLog,
	addTemplateLog,
	patchTemplateLog,
	patchIntent,
	getSpecific,
	analyzeEmail,
	smallTalk
} from "@functions/index";
import { default as connect } from "@db/connectMongo";

connect();

const serverlessConfiguration: AWS = {
	service: "stan-backend",
	frameworkVersion: "2",
	custom: {
	},
	plugins: [],
	provider: {
		name: "aws",
		runtime: "nodejs14.x",
		apiGateway: {
			minimumCompressionSize: 1024,
			shouldStartNameWithService: true,
		},
		environment: {
			AWS_NODEJS_CONNECTION_REUSE_ENABLED: "1",
		},
		lambdaHashingVersion: "20201221",
	},
	// import the function via paths
	functions: {
		getUser,
		addUser,
		delUser,
		patchUser,
		loginUser,
		getUserByToken,
		getBuildings,
		addBuilding,
		patchBuilding,
		delBuilding,
		// getTemplates,
		addTemplate,
		delTemplate,
		patchTemplate,
		getUserLogs,
		getTemplateLogs,
		addUserLog,
		addTemplateLog,
		patchTemplateLog,
		patchIntent,
		getSpecific,
		analyzeEmail,
		smallTalk
	},
};

module.exports = serverlessConfiguration;

```

</details>

<details>
<summary><b><code>SLS_DEBUG=* serverless deploy</code> output</b></summary>

```
Serverless: Load command interactiveCli
Serverless: Load command config
Serverless: Load command config:credentials
Serverless: Load command config:tabcompletion
Serverless: Load command config:tabcompletion:install
Serverless: Load command config:tabcompletion:uninstall
Serverless: Load command create
Serverless: Load command install
Serverless: Load command package
Serverless: Load command deploy
Serverless: Load command deploy:function
Serverless: Load command deploy:list
Serverless: Load command deploy:list:functions
Serverless: Load command invoke
Serverless: Load command invoke:local
Serverless: Load command info
Serverless: Load command logs
Serverless: Load command metrics
Serverless: Load command print
Serverless: Load command remove
Serverless: Load command rollback
Serverless: Load command rollback:function
Serverless: Load command slstats
Serverless: Load command plugin
Serverless: Load command plugin
Serverless: Load command plugin:install
Serverless: Load command plugin
Serverless: Load command plugin:uninstall
Serverless: Load command plugin
Serverless: Load command plugin:list
Serverless: Load command plugin
Serverless: Load command plugin:search
Serverless: Load command config
Serverless: Load command config:credentials
Serverless: Load command upgrade
Serverless: Load command uninstall
Serverless: Load command login
Serverless: Load command logout
Serverless: Load command generate-event
Serverless: Load command test
Serverless: Load command dashboard
Serverless: Load command output
Serverless: Load command output:get
Serverless: Load command output:list
Serverless: Load command param
Serverless: Load command param:get
Serverless: Load command param:list
Serverless: Load command studio
Serverless: Skipping variables resolution with old resolver (new resolver reported no more variables to resolve)
Serverless: Deprecation warning: Starting with next major version, "http.request.schema" property will be replaced by "http.request.schemas".
            More Info: https://www.serverless.com/framework/docs/deprecations/#AWS_API_GATEWAY_SCHEMAS
Serverless: Invoke deploy
Serverless: Invoke package
Serverless: Invoke aws:common:validate
Serverless: Invoke aws:common:cleanupTempDir
Serverless: Packaging service...
Serverless: Excluding development dependencies...
Db Connected!!
Serverless: Invoke aws:package:finalize
Serverless: Invoke aws:common:moveArtifactsToPackage
Serverless: Invoke aws:common:validate
Serverless: Invoke aws:deploy:deploy
Serverless: Uploading CloudFormation file to S3...
Serverless: Uploading artifacts...
Serverless: Uploading service stan-backend.zip file to S3 (36.22 MB)...
Serverless: Validating template...
Serverless: Updating Stack...
Serverless: Checking Stack update progress...
................................................................................................................................................
Serverless: Stack update finished...
Serverless: Invoke aws:info
Service Information
service: stan-backend
stage: dev
region: us-east-1
stack: stan-backend-dev
resources: 150
api keys:
  None
endpoints:
  GET - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/user/{email}
  POST - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/user
  DELETE - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/user
  PATCH - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/user
  POST - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/user/login
  GET - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/user/token/{token}
  GET - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/buildings/{building_token}
  POST - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/buildings
  PATCH - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/buildings
  DELETE - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/buildings
  POST - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/templates
  DELETE - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/templates
  PATCH - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/templates
  GET - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/logs/users
  GET - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/logs/templates
  POST - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/logs/users
  POST - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/logs/templates
  POST - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/logs/templates/review
  PATCH - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/intent
  GET - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/templates/{id}/{building_name}/{datetime}
  POST - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/user/email
  POST - https://rkmg39eisf.execute-api.us-east-1.amazonaws.com/dev/user/smalltalk
functions:
  getUser: stan-backend-dev-getUser
  addUser: stan-backend-dev-addUser
  delUser: stan-backend-dev-delUser
  patchUser: stan-backend-dev-patchUser
  loginUser: stan-backend-dev-loginUser
  getUserByToken: stan-backend-dev-getUserByToken
  getBuildings: stan-backend-dev-getBuildings
  addBuilding: stan-backend-dev-addBuilding
  patchBuilding: stan-backend-dev-patchBuilding
  delBuilding: stan-backend-dev-delBuilding
  addTemplate: stan-backend-dev-addTemplate
  delTemplate: stan-backend-dev-delTemplate
  patchTemplate: stan-backend-dev-patchTemplate
  getUserLogs: stan-backend-dev-getUserLogs
  getTemplateLogs: stan-backend-dev-getTemplateLogs
  addUserLog: stan-backend-dev-addUserLog
  addTemplateLog: stan-backend-dev-addTemplateLog
  patchTemplateLog: stan-backend-dev-patchTemplateLog
  patchIntent: stan-backend-dev-patchIntent
  getSpecific: stan-backend-dev-getSpecific
  analyzeEmail: stan-backend-dev-analyzeEmail
  smallTalk: stan-backend-dev-smallTalk
layers:
  None
Serverless: Invoke aws:deploy:finalize

```

</details>

<b>Installed version</b>

```
Framework Core: 2.44.0
Plugin: 5.2.0
SDK: 4.2.3
Components: 3.11.0
```

**Comment by pgrzesik** [comment]
Could you please clarify what do you mean that it `gets stuck`? Does it hang your terminal?

**Comment by premraval-pr** [comment]
Yes sure, sorry for not explaining completely. `gets stuck` means it hangs yes. If you see the SLS_DEGUG=* output, after the below line...

```
layers:
  None
Serverless: Invoke aws:deploy:finalize
```

It just stays here forever, no error shown or anything. I have tried the serverless deploy multiple times and waited for times ranging from 1 minute to 1 hour but every time it reached the layers: None and then stops.

**Comment by pgrzesik** [comment]
Thanks for additional clarification. What OS are you using? 

> It just stays here forever, no error shown or anything.

This is actually the last step - just to clarify 100% - after that line prints, your terminal fully hangs and you have to Ctrl+C to kill the process, right? Or you can run another commands afterwards?

**Comment by premraval-pr** [comment]
> Thanks for additional clarification. What OS are you using?

Windows 10 Pro

![image](https://user-images.githubusercontent.com/50078718/121182038-abb9d480-c830-11eb-917e-2fa2737d09e1.png)


> This is actually the last step - just to clarify 100% - after that line prints, your terminal fully hangs and you have to Ctrl+C to kill the process, right? Or you can run another commands afterwards?

Yes exactly, I have to kill the process (Ctrl + C) to do anything else. No, I can not run any other commands.

**Comment by premraval-pr** [comment]
UPDATE: I have solved it, it was 2 things that were bugging

1. I called a connectMongo() method in serverless.ts which opened a connection to MongoDB Atlas using mongoose API, which worked fine in `serverless offline`  because in the offline version we do open a localhost server, but as soon as it goes to serverless deploy it still starts that mongo connection while deploying and keeps it on forever. 

    **Solution**: After some research, I found out, that the correct way to use MongoDB with Lamda is to create separate connections in the lambda functions, hence removing the global conectMongo() function.

2. I was using `bcrypt` library to encrypt the passwords and comparing hash while login. And after exploring documentation and some blogs I found out that bcrypt gives error on Lambda and throws `/var/task/node_modules/bcrypt/lib/binding/napi-v3/bcrypt_lib.node: invalid ELF header`.

    **Solution**: I replaced `bcrypt` to `bcryptjs` and all of it works fine.

I am closing this issue as `serverless deploy` worked now and the API Endpoints are working.  I just added the errors above that  I faced from the `aws-nodejs-typescript` template so people can see and get some help if they face similar issues.
